_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+
+Minimal access to target environments::
+  All the infrastructure is deployed from within the accounts using a serverless platform. This means that attackers
+  that gain access to the underlying AWS secrets used by the CI environments will at most have the ability to run
+  deployments against a predefined set of code. This means that external attackers who do not have access to the source
+  code will at most be able to: (a) deploy code that has already been deployed before, (b) see the plan of the
+  infrastructure between two points of time. They will not be able to write arbitrary infrastructure code to read DB
+  secrets, for example. It is important to note that the IAM policies are set up such that the IAM user for CI only has
+  access to trigger predefined events. They do not have access to arbitrarily invoke the ECS task, as that could
+  potentially expose arbitrary deployments by modifying the command property (e.g use command to `echo` some
+  infrastructure code and run `terraform`).
+    - Note that there is still risk of rolling back the existing infrastructure by attempting to deploy a previous
+      version. See below for potential ways to mitigate this type of attack.
+    - Similarly, this alone does not mitigate threats from internal attackers who have access to the source code, as a
+      potential attacker with access to the source code can write arbitrary code to destroy or lookup arbitrary
+      infrastructure in the target environment. See below for potential ways to mitigate this type of attack.
+
+Minimal options for deployment::
+  The Lambda function exposes a minimal interface for triggering deployments. Attackers will only be able to trigger a
+  deployment against a known repo and known git refs (branches, tags, etc). To further limit the scope, the lambda
+  function can be restricted to only allow references to repositories that matches a predefined regular expression.
+  Terraform Enterprise exposes similar configuration parameters to restrict what deployments can be triggered. This
+  prevents attackers from creating an open source repo with malicious code that they subsequently deploy by pointing the
+  deploy runner to it.
+
+Restricted refs for `apply`::
+  Since many CI systems depend on the pipeline being managed as code in the same repository, internal attackers can
+  easily circumvent approval flows by modifying the CI configuration on a test branch. This means that potential
+  attackers can run an `apply` to destroy the environment or open backdoors by running infrastructure code from test
+  branches without having the code approved. To mitigate this, the Lambda function allows specifying a list of git refs
+  (branches, tags, etc) as the source of `apply` and `apply-all`. If you limit the source of `apply` to only protected
+  branches (see below), it prevents attackers from having the ability to run `apply` unless it has been reviewed.
+
+CI server does not need access to the source code::
```

Comment:
Big +1 on this! In fact... We should make sure we set up our Ref Arch and our own dogfood Ref Arch exactly this way: no write access to infra live / infra modules. This does mean rethinking how we do CI / CD for apps (which today _does_ write to infra-live).

Reply:
There is one problem with this that I have not yet quite figured out, which is that in order to use `git-updated-folders`, you need access to the source code.

I think in the future, we should push this feature down to the deploy script so that we can achieve this promise, but not quite sure yet what that interface looks like. Perhaps something like `--run-on-changed-files` and `--terragrunt-changes`, `--terraform-changes`, `--file-changes FILE_REGEX`?

Reply:
> There is one problem with this that I have not yet quite figured out, which is that in order to use git-updated-folders, you need access to the source code.

Read access is enough though, right? My point is that the CI server shouldn't need write access.